light-client/src/light_client.rs

```diff
@@ -230,35 +232,32 @@ impl LightClient {
         }
     }
 
-    /// Look in the light store for a block from the given peer at the given height.
+    /// Look in the light store for a block from the given peer at the given height,
+    /// whatever its verification status.
+    ///
     /// If one cannot be found, fetch the block from the given peer.
     ///
     /// ## Postcondition
     /// - The provider of block that is returned matches the given peer.
-    #[post(ret.as_ref().map(|lb| lb.provider == self.peer).unwrap_or(true))]
+    #[post(ret.as_ref().map(|(lb, _)| lb.provider == self.peer).unwrap_or(true))]
     pub fn get_or_fetch_block(
         &self,
-        current_height: Height,
+        height: Height,
         state: &mut State,
-    ) -> Result<LightBlock, Error> {
-        let current_block = state
-            .light_store
-            .get(current_height, Status::Verified)
-            .or_else(|| state.light_store.get(current_height, Status::Unverified));
+    ) -> Result<(LightBlock, Status), Error> {
+        let block = state.light_store.get_non_failed(height);
 
-        if let Some(current_block) = current_block {
-            return Ok(current_block);
+        if let Some(block) = block {
+            return Ok(block);
         }
 
-        self.io
-            .fetch_light_block(self.peer, AtHeight::At(current_height))
-            .map(|current_block| {
-                state
-                    .light_store
-                    .insert(current_block.clone(), Status::Unverified);
+        let block = self
+            .io
+            .fetch_light_block(self.peer, AtHeight::At(height))
+            .map_err(ErrorKind::Io)?;
+
+        state.light_store.insert(block.clone(), Status::Unverified);
 
-                current_block
-            })
-            .map_err(|e| ErrorKind::Io(e).into())
+        Ok((block, Status::Unverified))
```

Comment:
Encoding this information as tuples is surprising, esp. as we construct permutations of it on the spot. Can't we more strongly encode this in an enum, which for some variants carries the `LightBlock`?

Reply:
This is in line with our discussions about reasoning about the state of the light client and how we should thrive for stronger abstractions which make it hard to construct impossible states.

Reply:
> Encoding this information as tuples is surprising, esp. as we construct permutations of it on the spot.

To me a tuple is basically an anonymous struct, and I figured since the status of a block acts like a key, it would be okay to package both the block and its status as an enum, much like part of the API of Rust HashMap etc. where they return the value and its key as a tuple, rather than making up a struct whose only purpose is to be deconstructed at call site.

> Can't we more strongly encode this in an enum, which for some variants carries the LightBlock?

I don't quite see how an enum would fit here, could you expand on that?

Reply:
I think this goes in line with https://github.com/informalsystems/tendermint-rs/pull/419#issuecomment-656591438 and the general discussion around representing states in the crate.

Specifically here I'm wondering are we ever interested in a `LightBlock` without the status information? How would the rest of the crate look like if this would be our `LightBlock`:

```rust
enum LightBlock {
    Failed {
        provider: PeerId,
    },
    Unverified {
        provider: PeerId,
    },
    Verified {
        signed_header: SignedHeader,
        validators: ValidatorSet,
        next_validators: ValidatorSet,
        provider: PeerId,
    },
    Trusted {
        signed_header: SignedHeader,
        validators: ValidatorSet,
        next_validators: ValidatorSet,
        provider: PeerId,
    },
}
```

Generally what I think we see here is that we let the actual domain of block storage creep into our higher level abstractions, which is informed too much by the key/value store semantics and it's physicality.

Reply:
> Generally what I think we see here is that we let the actual domain of block storage creep into our higher level abstractions, which is informed too much by the key/value store semantics and it's physicality.

Agree 100%. We should think hard about this as part of https://github.com/informalsystems/tendermint-rs/issues/428.

Reply:
Absolutely, let's move the discussion there!

Reply:
I am not quite convinced that the enum above is the right way to express this, but I agree that we should look into this more closely.